File: cpp/prime.cpp
#include <iostream>

#define NULL 0

class Sieve {
private:
    int _prime;
    Sieve* _next = NULL;

public:
    Sieve(int prime) {
        _prime = prime;
    }

    void test(int n) {
        if (_prime * (n / _prime) == n) {
        } else if (_next == NULL) {
            std::cout << n << std::endl;
            _next = new Sieve(n);
        } else {
            _next->test(n);
        }
    }
};

int main() {
    Sieve* s = new Sieve(2);
    int i = 2;

    std::cout << "2" << std::endl;

    while (i < 12) {
        i += 1;
        s->test(i);
    }

    return 0;
}

File: typescript/prime.ts
class Sieve {

  private _prime: number;
  private _next: Sieve;

  constructor(prime: number) {
    this._prime = prime;
  }

  public test(n: number) {
    if ((this._prime * (n / this._prime)) == n) {
    }
    else if ((this._next == null)) {
      console.log(n);
      this._next = new Sieve(n);
    }
    else {
      this._next.test(n);
    }

  }
}

function main() {
  let s = new Sieve(2);
  let i = 2;
  console.log("2");
  while ((i < 10)) {
    i++;
    s.test(i);
  }
}

main();